File: web/lib/screener.ts
export type SignalKind = "entry" | "exit" | "none";

export interface ScreenerRow {
  symbol: string;
  name: string;
  price: number;
  change_percent: number;
  volume: number;
  value: number;
  inflow: number;
  outflow: number;
  net_flow: number;
  buy_volume: number;
  sell_volume: number;
  buy_ratio: number | null;
  sell_ratio: number | null;
  volume_surge: number | null;
  score: number;
  entry_signal: boolean;
  exit_signal: boolean;
  unexpected: boolean;
  flow_verified: boolean;
  tracked: boolean;
  vwap: number | null;
  atr: number | null;
  bid: number | null;
  ask: number | null;
  book_pressure: number | null;
  suggested_entry: number | null;
  suggested_exit: number | null;
  target_price: number | null;
  stop_loss: number | null;
  reasons: string[];
  sources: string[];
  updated_at: string | null;
}

export interface MarketPulse {
  index: string;
  index_value: number | null;
  index_change_percent: number | null;
  advancing: number | null;
  declining: number | null;
  delayed: boolean;
}

export interface ScreenerSnapshot {
  watchlist: ScreenerRow[];
  radar: ScreenerRow[];
  pulse: MarketPulse;
  scanned: number;
  delayed: boolean;
  updated_at: string | null;
}

export function signalKind(row: Pick<ScreenerRow, "entry_signal" | "exit_signal">): SignalKind {
  if (row.entry_signal) return "entry";
  if (row.exit_signal) return "exit";
  return "none";
}

export function signalKey(row: ScreenerRow): string {
  return `${row.symbol}:${row.entry_signal ? "e" : ""}${row.exit_signal ? "x" : ""}:${row.suggested_entry ?? ""}:${row.suggested_exit ?? ""}`;
}

export function toFiniteNumber(value: unknown): number | null {
  if (value == null || value === "") return null;
  const parsed = typeof value === "number" ? value : Number(value);
  return Number.isFinite(parsed) ? parsed : null;
}

export function parseRow(raw: Record<string, unknown>): ScreenerRow {
  return {
    symbol: String(raw.symbol ?? ""),
    name: String(raw.name ?? ""),
    price: toFiniteNumber(raw.price) ?? 0,
    change_percent: toFiniteNumber(raw.change_percent) ?? 0,
    volume: toFiniteNumber(raw.volume) ?? 0,
    value: toFiniteNumber(raw.value) ?? 0,
    inflow: toFiniteNumber(raw.inflow) ?? 0,
    outflow: toFiniteNumber(raw.outflow) ?? 0,
    net_flow: toFiniteNumber(raw.net_flow) ?? 0,
    buy_volume: toFiniteNumber(raw.buy_volume) ?? 0,
    sell_volume: toFiniteNumber(raw.sell_volume) ?? 0,
    buy_ratio: toFiniteNumber(raw.buy_ratio),
    sell_ratio: toFiniteNumber(raw.sell_ratio),
    volume_surge: toFiniteNumber(raw.volume_surge),
    score: toFiniteNumber(raw.score) ?? 0,
    entry_signal: Boolean(raw.entry_signal),
    exit_signal: Boolean(raw.exit_signal),
    unexpected: Boolean(raw.unexpected),
    flow_verified: Boolean(raw.flow_verified),
    tracked: Boolean(raw.tracked),
    vwap: toFiniteNumber(raw.vwap),
    atr: toFiniteNumber(raw.atr),
    bid: toFiniteNumber(raw.bid),
    ask: toFiniteNumber(raw.ask),
    book_pressure: toFiniteNumber(raw.book_pressure),
    suggested_entry: toFiniteNumber(raw.suggested_entry),
    suggested_exit: toFiniteNumber(raw.suggested_exit),
    target_price: toFiniteNumber(raw.target_price),
    stop_loss: toFiniteNumber(raw.stop_loss),
    reasons: Array.isArray(raw.reasons) ? raw.reasons.map(String) : [],
    sources: Array.isArray(raw.sources) ? raw.sources.map(String) : [],
    updated_at: raw.updated_at == null ? null : String(raw.updated_at),
  };
}

File: app/models/screener.py
from __future__ import annotations

import re
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, field_validator

_ARABIC_DIGITS = str.maketrans("٠١٢٣٤٥٦٧٨٩", "0123456789")
_SYMBOL_RE = re.compile(r"^\d{4}$")
_MAIN_MARKET_RE = re.compile(r"^[1-8]\d{3}$")


def normalize_tasi_symbol(value: str) -> str:
    symbol = value.strip().translate(_ARABIC_DIGITS).upper()
    if not _SYMBOL_RE.fullmatch(symbol):
        raise ValueError(symbol)
    return symbol


def is_tasi_main_symbol(value: str) -> bool:
    """True for Tadawul main-market equities (1xxx–8xxx), not Nomu/ETF 9xxx."""

    ticker = str(value or "").strip().translate(_ARABIC_DIGITS).upper()
    return bool(_MAIN_MARKET_RE.fullmatch(ticker))


class WatchlistItemIn(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    symbol: str = Field(..., min_length=1, max_length=8)

    @field_validator("symbol")
    @classmethod
    def normalize(cls, value: str) -> str:
        try:
            return normalize_tasi_symbol(value)
        except ValueError as exc:
            raise ValueError("استخدم رمز تداول من 4 أرقام مثل 4030") from exc


class WatchlistResponse(BaseModel):
    symbols: list[str]
    count: int


class ScreenerRow(BaseModel):
    symbol: str
    name: str = ""
    price: Decimal = Decimal("0")
    change_percent: Decimal = Decimal("0")
    volume: Decimal = Decimal("0")
    value: Decimal = Decimal("0")
    inflow: Decimal = Decimal("0")
    outflow: Decimal = Decimal("0")
    net_flow: Decimal = Decimal("0")
    buy_volume: Decimal = Decimal("0")
    sell_volume: Decimal = Decimal("0")
    buy_ratio: Decimal | None = None
    sell_ratio: Decimal | None = None
    volume_surge: Decimal | None = None
    score: Decimal = Decimal("0")
    entry_signal: bool = False
    exit_signal: bool = False
    unexpected: bool = False
    flow_verified: bool = False
    tracked: bool = False
    vwap: Decimal | None = None
    atr: Decimal | None = None
    bid: Decimal | None = None
    ask: Decimal | None = None
    book_pressure: Decimal | None = None
    suggested_entry: Decimal | None = None
    suggested_exit: Decimal | None = None
    target_price: Decimal | None = None
    stop_loss: Decimal | None = None
    reasons: list[str] = Field(default_factory=list)
    sources: list[str] = Field(default_factory=list)
    updated_at: datetime | None = None


class MarketPulse(BaseModel):
    index: str = "TASI"
    index_value: Decimal | None = None
    index_change_percent: Decimal | None = None
    advancing: int | None = None
    declining: int | None = None
    delayed: bool = True


class ScreenerSnapshot(BaseModel):
    watchlist: list[ScreenerRow]
    radar: list[ScreenerRow]
    pulse: MarketPulse
    scanned: int
    delayed: bool = True
    updated_at: datetime | None = None
